### ISSUE TITLE
Support MDF/MDS CDROM image files

### ISSUE THREAD

**Issue by weirddan455** [dosbox-staging/dosbox-staging#3805]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

None

### Is your feature request related to a problem? Please describe.

These are formats used by Alcohol 120%.  Would be nice to support them.  MDS is a small file that does the same job as a CUE file, only it's encoded in a binary format rather than text.  MDF is the BIN equivalent.  If we're lucky, all we would need to do is write some parsing code for the small MDS file to read track information and it can use the existing code for BIN/CUE.

### Describe the solution you'd like

Write an MDS parser and see if it works.

### Describe alternatives you've considered

Maybe there's existing code from another project we can use.

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
MDF/MDS can be used to store images that use disk-based protection. Not sure if X handles that or just reads the pure user data part.

**Comment by weirddan455** [comment]
Yeah, I think that's correct @johnnovak 

We can probably get away without handling copy protections though.  I know a few DOS games use LaserLock.  Not sure if that one needs MDF/MDS or if BIN/CUE is good enough but we don't handle that currently.

IMO just getting this file format working at all for non-copy protected games would be a nice improvement.

**Comment by johnnovak** [comment]
> IMO just getting this file format working at all for non-copy protected games would be a nice improvement.

Fair point, and probably all we need.

And just pointing out that supporting Win98 "officially" would open the floodgates in this area, too. Many Win9x-era games use disk-based protections.

**Comment by weirddan455** [comment]
>  And just pointing out that supporting Win98 "officially" would open the floodgates in this area, too.

Are you planning on doing that?

We can look into implementing copy protections later as it would benefit a small minority of DOS games as well.  That's a separate task in my mind.  LaserLock is a more primitive one that appears to use a hidden directory.  It probably needs some tinkering on the MSCDEX layer to get that working.  Should be do-able but I haven't looked into it thoroughly.

The Win9x era had some more advanced copy protection schemes.  I vaguely remember they worked based off some timings due to "wobbles" on the disk or something.  I also remember using Alcohol 120% and managing to burn a copy of some of my games that worked without a no-CD crack so that's also possible, but harder.

**Comment by Grounded0** [comment]
It would nice to have a list of games with optical disc copy protection scheme for DOS, Win16 and Win32s.

I think John Novak just pointed out that IF we were to officially support Win9x this would become more important. Can't see that happening since i don't see many motivated people here eager to solve issues with [Monster Truck Madness (1996)](https://www.mobygames.com/game/2584/monster-truck-madness/).

**Comment by johnnovak** [comment]
> I don't want to see this issue tracker turning into a dumping ground where people have very little motivation in actually resolving people's problems. The bigger the issue tracker gets less valuable it becomes as a means of project coordination tool.

I agree with that, in an abstract sense.

But we're in no danger of that at present, our issue tracker is in a good shape, and @weirddan455 is interested in working on adding support for this.

**Comment by Grounded0** [comment]
Yeah i agree that this is a good feature to add. I've seen MDF/MDS disc image dumps on archive.org for DOS/Win 3.x stuff that i couldn't use but had instead hunt for an ISO.

**Comment by FeralChild64** [comment]
Here is another open-source project, supporting MDS format: https://cdemu.sourceforge.io/

**Comment by MasterO2** [comment]
> Here is another open-source project, supporting MDS format: https://cdemu.sourceforge.io/

Mirrored on Github as well: https://github.com/cdemu/cdemu

**Comment by Grounded0** [comment]
> But we're in no danger of that at present, our issue tracker is in a good shape, and @weirddan455 is interested in working on adding support for this.

Yeah I agree. This was directed at us hypothetically introducing an official Windows 9x support which almost everyone agrees is a bad idea for now.

My schizoaffective disorder has been getting much worse and I need to dial back my activities here. Hopefully nobody misinterpreted anything here as negative statements directed at them. 🙏

**Comment by johnnovak** [comment]
@Grounded0 It's all good man, and I wish you all the best ❤️

**Comment by weirddan455** [comment]
@Grounded0 No worries, I didn't take offense to anything you said.  I hope you feel better soon.

> [X supports MDF](https://dosbox-x.com/wiki/#Guide:Managing-image-files-in-DOSBox%E2%80%90X)

I looked at what DOSBox-X was doing and they're not really supporting it properly.  I was able to mount a .MDF image in X that doesn't mount in Staging but the .MDS file does not mount in either.  They've got some different logic in their `LoadIsoFile` and it happened to load it as though it were an ISO file.  This is likely to fail with images containing audio tracks because X has no parser for MDS.

> > Here is another open-source project, supporting MDS format: https://cdemu.sourceforge.io/
> 
> Mirrored on Github as well: https://github.com/cdemu/cdemu

I am familiar with the cdemu projecting, having used it myself for testing the physical CDROM implementation in Staging.  They have a library called libMirage for parsing different image formats but unfortunately it has a dependency on glib.  I really don't want anything else in Staging depending on glib as it has been problematic in our build system.

I might end end up using their code as a reference though.  This is their MDS parser:  https://github.com/cdemu/cdemu/blob/master/libmirage/images/image-mds/parser.c

I'd prefer to find documentation on the MDS specification but I've come up empty on that so far.  It seems to be proprietary and cdemu probably had to reverse engineer it.

**Comment by FeralChild64** [comment]
There is probably no official specification anywhere, just some reverse engineered documentation:

https://www.cryer.co.uk/file-types/m/mds/mds.txt
http://forum.redump.org/post/41803/#p41803

**Comment by weirddan455** [comment]
@FeralChild64 Better than nothing!  Nice find.

**Comment by TaraLonger** [comment]
mdf support is should not a high priority or implemented at all. Any MDf that i encounter in dosbox x i usually convert to cue and bin because dosbox only mounts them as data without any of the extract stuff. and if they are multitrack it only mounts the first track.  I have found that once they are converted from mdf to cue and bin using ultra iso they actually work better.

**Comment by johnnovak** [comment]
Thanks @TaraLonger , that's very useful info. Added the low-priority label.

**Comment by weirddan455** [comment]
Adding this to 0.83.0 milestone. We don't need to hold up the release if this isn't finished but I think I can get it done in time.
